src/core.rs

```diff
@@ -542,18 +563,74 @@ where
                     ));
                 }
                 GenericAction::Union(span, e1, e2) => {
-                    let mapped_e1 =
-                        e1.to_core_actions(typeinfo, binding, fresh_gen, &mut norm_actions)?;
-                    let mapped_e2 =
-                        e2.to_core_actions(typeinfo, binding, fresh_gen, &mut norm_actions)?;
-                    norm_actions.push(GenericCoreAction::Union(
-                        span.clone(),
-                        mapped_e1.get_corresponding_var_or_lit(typeinfo),
-                        mapped_e2.get_corresponding_var_or_lit(typeinfo),
-                    ));
-                    mapped_actions
-                        .0
-                        .push(GenericAction::Union(span.clone(), mapped_e1, mapped_e2));
+                    match (e1, e2) {
```

Comment:
Does it work both ways?
```
(GenericExpr::Var(..), GenericExpr::Call(_, f, args)) | (GenericExpr::Call(_, f, args), GenericExpr::Var(..))
                            if f.is_func(typeinfo)
```

Reply:
Do you also need to check that `f` is a constructor? E.g., consider a function with(out) a merge function
```
(datatype E (g) (h))
(function f () E :no-merge)
(set (f) (g))


(union (f) (h))
;; is different from
;; (set (f) (h))

```

Setting and unioning have different semantics.

Reply:
Made both changes!